### ISSUE TITLE
Just use BroadcastArray from LazyArrays?

### ISSUE THREAD

**Issue by dlfivefifty** [JuliaArrays/GetindexArrays.jl#3]
Note you can accomplish the same thing via, eg, 
```Julia
f = (k,j) -> …
BroadcastArray(f, 1:n, (1:m)’)
```
This supports arbitrary axes including offset.

**Comment by timholy** [comment]
I knew something like this had to exist. That said, it looks like it's not a subtype of AbstractArray. That's going to be a compatibility problem for *many* uses.

**Comment by dlfivefifty** [comment]
Of course it’s a subtype of ` AbstractArray `

**Comment by dlfivefifty** [comment]
Note also it supports parameters so avoids closure issues:
```julia
BroadcastVector((k,x) -> cos(k*x), 1:5, 0.1)
```

**Comment by timholy** [comment]
Oh, I misunderstood the meaning of the README (it was referring to `Broadcasted`, not `BroadcastArray`). 

I'm noticing a performance issue, though. Modifying https://github.com/JuliaArrays/GetindexArrays.jl/blob/45cae430bfd16fdfe7651fea3b82b25afee3447b/test/runtests.jl#L74-L107

to use `BroadcastArray`, which requires something like
```julia
using ImageFiltering   # for kernelfactors (adjoint only covers up to 2d, this is 3d so higher-dimensional axis reshaping is needed)
function make_all(P, tforms)
    # forced specialization
    capture_getindex(i::Vararg{Int,N}) where {N} = special_getindex(P, tforms, i...)
    A = @inferred(BroadcastArray(capture_getindex, kernelfactors(axes(P))...))
    return A
end
```
Then I get
```julia
julia> using BenchmarkTools

julia> @btime $A[i,j,k] setup=(i=rand(axes(A,1)); j=rand(axes(A,2)); k=rand(axes(A,3)))
  178.280 ns (14 allocations: 480 bytes)
0.0
```
whereas with this package it's
```julia
julia> @btime $A[i,j,k] setup=(i=rand(axes(A,1)); j=rand(axes(A,2)); k=rand(axes(A,3)))
  113.538 ns (11 allocations: 352 bytes)
```

Update: SimpleTraits has a performance issue; if I stop using it in ImageAxes and just handle the traits manually, then I get
about 48ms for BroadcastArray and 26ms for this package. So it's almost a factor of 2 difference. That's enough for me to care about, though if BroadcastArray can be fixed then I'd be happy to delete this repo.

Thoughts?

**Comment by dlfivefifty** [comment]
I didn't try your explicit example but note that `getindex` at a single index is going to be hit by overhead by checking for special broadcasting due to 1-length. Note it's reducing to `getindex(::Base.Broadcasting.Broadcast, k::Integer, j::Integer)`  so any slowness is due to slowness in broadcasting. It might be possible to special case here where at compile time we can recognise the dimensions which are 1.

Note LazyArrays.jl does _much_ better with slices than with single `getindex` calls. If you take a slice they are around the same time:
```julia
julia> n = 1000; A = GetindexArray((k,j) -> cos(k+j), (Base.OneTo(n), Base.OneTo(n)))

julia> @btime A[1:500,1:500];
  2.661 ms (2 allocations: 1.91 MiB)

julia> n = 1000; A = BroadcastArray((k,j) -> cos(k+j), Base.OneTo(n), Base.OneTo(n)')

julia> @btime A[1:500,1:500];
  2.652 ms (2 allocations: 1.91 MiB)
```
And  if you want parameters it's a 4x performance improvement usng `BroadcastArray`:
```julia
julia> let x = 0.1
       n = 1000; A = GetindexArray((k,j) -> cos(x * (k+j)), (Base.OneTo(n), Base.OneTo(n)));
       end

julia> @btime A[1:500,1:500];
  14.769 ms (619807 allocations: 11.36 MiB)

julia> x = 0.1; n = 1000; A = BroadcastArray((k,j,x) -> cos(x*(k+j)), Base.OneTo(n), Base.OneTo(n)',x);

julia> @btime A[1:500,1:500];
  3.040 ms (2 allocations: 1.91 MiB)
```

**Comment by timholy** [comment]
Yep, though #1 does the same parameters trick.

This package was inspired by a demo where I'm making the point that scalar-indexing performance matters, so I don't want to sweep it under the rug. And it's possible to have no-overhead scalar indexing here.

**Comment by timholy** [comment]
Hah, we can of course do the same thing with MappedArrays, and there it seems there's no performance penalty. Construction is with `mappedarray(f, CartesianIndices(axs))`. Maybe we really don't need this package.

**Comment by dlfivefifty** [comment]
Your `Extras` seems like it's trying to recreate broadcasting, just without the 1-length special case...

Ideally in Julia v2.0 the following changes:
1. Scalars have special axes, say called `Scalar` (or `OneToOne`)
2. transpose/adjoints have `axes(a', 1) isa Scalar `
3. Broadcasting which propagates along other axes is only supported when the axes is either past the axes or is `Scalar`.

This would mean special cases can be done at compile-time with no performance penalty. It would also make clearer that `a'*b` returns a `Number`.

**Comment by timholy** [comment]
I totally agree that changes to broadcasting have wide-ranging benefits when there's an unambiguous improvement, and if some of those can't be done in Julia 1.x because they would be breaking, then we should plan on fixing what we can for Julia 2.0.

But for this particular effort, I'm actually not focused on broadcasting; it's more about supporting things like spatial deformations of the underlying grid. For example, image registration of a timeseries 2d or 3d movie, where the original array might be many TB. JuliaImages & JuliaArrays support a ton of lazy operations in composable fashion, but layering has its disadvantages sometimes, and so fusing them together into a single closure does have advantages.
